From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shyam Jeedigunta <jeedigv@amazon.com>
Date: Wed, 7 Sep 2022 13:23:53 -0700
Subject: [PATCH] --EKS-PATCH-- Reduce default gzip compression level from 4 to
 1 in apiserver

Cherry-pick of upstream Kubernetes PR #112299 (https://github.com/kubernetes/kubernetes/pull/112299),
which has been merged into upstream Kubernetes and back-ported to all minor version branches >= 1.23.
For versions <= 1.22 which Kubernetes no longer intends to release new patch versions for, we are
back-porting the patch.

From the original PR description:
    kube-apiserver: gzip compression switched from level 4 to level 1 to improve large list call
    latencies in exchange for higher network bandwidth usage (10-50% higher). This increases the
    headroom before very large unpaged list calls exceed request timeout limits.
---
 .../pkg/endpoints/handlers/responsewriters/writers.go       | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go b/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go
index 16e16c35376..2a50c1c589c 100644
--- a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go
+++ b/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go
@@ -143,8 +143,10 @@ var gzipPool = &sync.Pool{
 }
 
 const (
-	// defaultGzipContentEncodingLevel is set to 4 which uses less CPU than the default level
-	defaultGzipContentEncodingLevel = 4
+	// defaultGzipContentEncodingLevel is set to 1 which uses least CPU compared to higher levels, yet offers
+	// similar compression ratios (off by at most 1.5x, but typically within 1.1x-1.3x). For further details see -
+	// https://github.com/kubernetes/kubernetes/issues/112296
+	defaultGzipContentEncodingLevel = 1
 	// defaultGzipThresholdBytes is compared to the size of the first write from the stream
 	// (usually the entire object), and if the size is smaller no gzipping will be performed
 	// if the client requests it.
